histomicstk/preprocessing/color_deconvolution/color_deconvolution.py

```diff
@@ -65,27 +74,195 @@ def color_deconvolution(im_rgb, w, I_0=None):
+    .. [#] Xu, J., Xiang, L., Wang, G., Ganesan, S., Feldman, M., Shih, N. N.,
+           ...& Madabhushi, A. (2015). Sparse Non-negative Matrix Factorization
+           (SNMF) based color unmixing for breast histopathological image
+           analysis.  Computerized Medical Imaging and Graphics, 46, 20-29.
+
+    """
+    stain_unmixing_method = stain_unmixing_method.lower()
+
+    if stain_unmixing_method == 'macenko_pca':
+        stain_deconvolution = rgb_separate_stains_macenko_pca
+        stain_unmixing_params['I_0'] = None
+        stain_unmixing_params['mask_out'] = mask_out
+
+    elif stain_unmixing_method == 'xu_snmf':
+        stain_deconvolution = rgb_separate_stains_xu_snmf
+        stain_unmixing_params['I_0'] = None
+        assert mask_out is None, "Masking is not yet implemented in xu_snmf."
+
+    else:
+        raise ValueError("Unknown/Unimplemented deconvolution method.")
+
+    # get W_source
+    W_source = stain_deconvolution(im_rgb, **stain_unmixing_params)
+
+    # If Macenco method, reorder channels in W_target and W_source as desired.
+    # This is actually a necessary step in macenko's method since we're
+    # not guaranteed the order of the different stains.
+    if stain_unmixing_method == 'macenko_pca':
+        W_source = _reorder_stains(W_source, stains=stains)
+
+    return W_source
+
+
+def color_deconvolution_routine(
+        im_rgb, W_source=None, mask_out=None, **kwargs):
+    """Unmix stains mixing followed by deconvolution (wrapper).
+
+    Parameters
+    ------------
+    im_rgb : array_like
+        An RGB image (m x n x 3) to colro normalize
+
+    W_source : np array, default is None
+        A 3x3 matrix of source stain column vectors. Only provide this
+        if you know the stains matrix in advance (unlikely) and would
+        like to perform supervised deconvolution. If this is not provided,
+        stain_unmixing_routine() is used to estimate W_source.
+
+    mask_out : array_like, default is None
+        if not None, should be (m x n) boolean numpy array.
+        This parameter ensures exclusion of non-masked areas from calculations
+        and stain matrix. This is relevant because elements like blood,
+        sharpie marker, white space, cannot be modeled as a mix of two stains.
+
+    kwargs : k,v pairs
+        Passed as-is to stain_unmixing_routine() if W_source is None.
+
+    Returns
+    --------
+    Output from color_deconvolution()
+
+    See Also
+    --------
+    histomicstk.preprocessing.color_deconvolution.stain_unmixing_routine
+    histomicstk.preprocessing.color_deconvolution.color_deconvolution
+
+    """
+    # get W_source if not provided
+    if W_source is None:
+        W_source = stain_unmixing_routine(im_rgb, mask_out=mask_out, **kwargs)
+
+    # deconvolve
+    Stains, StainsFloat, wc = color_deconvolution(im_rgb, w=W_source, I_0=None)
+
+    # mask out (keep in mind, image is inverted)
+    if mask_out is not None:
+        for i in range(3):
```

Comment:
Instead of `range(3)`, would it be better to generalize this is a `range(Stains.shape[-1])`?  If so, I think `range(3)` is used elsewhere, too.

Reply:
Actually, the masking out is currently **only** implemented for 3 stains, so even if the matrix has two elements, it needs to be complemented into a 3-stain matrix . So having this explicitly shown this way helps emphasize this. In this future we may generalize, but this is the status of the current codebase.